Sources/TensorFlow/Operators/NN.swift

```diff
@@ -12,6 +12,44 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//===------------------------------------------------------------------------------------------===//
+// Core
+//===------------------------------------------------------------------------------------------===//
+
+private extension Tensor where Scalar: TensorFlowFloatingPoint {
+    /// Returns a dropout given a probability and noise shape.
+    @differentiable(wrt: self where Scalar: Differentiable)
+    func droppingOut(probability: Double, noiseShape: TensorShape) -> Tensor {
+        let noise = Tensor(randomUniform: noiseShape)
+        let keepMask = noise .>= Scalar(probability)
+        let keepProbability = Scalar(1.0 - probability)
+        return self * Tensor(keepMask) / Tensor(keepProbability)
+    }
+}
+
+/// Returns the result of applying dropout on an input tensor.
+///
+/// - Parameters:
+///     - input: The input units.
+///     - probability: The probability that a input unit will be dropped.
+///     - noiseShape: The shape of the randomly generated keep mask.
+/// - Precondition: `probability` is within the range 0..<1.
+/// - Note: `dropout` is an identity function on `input` when called during the inference phase of
+///     learning.
+@differentiable(wrt: input)
+public func dropout<Scalar: TensorFlowFloatingPoint>(
```

Comment:
Why did you also add a global function aside from `Tensor.droppingOut`? I believe we should not be mixing `Context`s in the operator definitions and stick to only using them within `Layer`s so we don't mix up high-level and low-level APIs. @rxwei what do you think?

Reply:
As you noted below, I restricted `Tensor.droppingOut` to `private` because it doesn't make sense, at least for me, to expose sort of half an implementation of `dropout` given that the scaling of samples in `Tensor.droppingOut` is useful only in the context of learning phases. I don't find `Tensor.droppingOut`, defined as inverted dropout, to be useful to the outside world given its purpose.

Now the purpose of abstracting out the `dropout` function with `Context` was to avoid repetitively switching on `learningPhase` for each of the four dropout layers' `callAsFunction`. It was just a reduction of ~24 lines. But I have no problem bringing that back into the layer and providing `Tensor.droppingOut` publicly if that's desired.

Reply:
@eaplatanios, @rxwei: 
I'm going to upgrade `Tensor.droppingOut(probability:noiseShape:)` to public in `NN.swift`. 

Should I move `dropout(_:probability:noiseShape:)` to `Layers/Core.swift` as a free function with private access or move its contents, the switch on learning phase, into each of the dropouts' `callAsFunction` just as it was previously handled in `Dropout`? 
Thanks!